### PR TITLE
Hatkiser/data fetching improvement

### PR DESCRIPTION
**Pull request by hatskier**

### CODE REVIEW

File: utils/v2/connector/impl/DataPackageUtils.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+  const sortedPackages = [...packages];
+  sortedPackages.sort((p1, p2) => p1.timestamp - p2.timestamp); // sorting prices from oldest to newest
+  switch (valueSelectionAlgorithm) {
+
+    // TODO: improve the implementation for first-valid
+    // We don't really need to wait for all responses within
+    // the first-valid method. It's enough to receive the first
+    // valid answer
+    case "first-valid":
+      return packages[packages.length - 1];
```
Comment (jakub-wojciechowski) [inline utils/v2/connector/impl/DataPackageUtils.ts:22]: Hmm, maybe I'm missing sth, but I don't see much added value with having "first-valid" vs "newest-valid" here.

File: utils/v2/connector/impl/DataPackageUtils.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  // Offchain signature verification
+  // (only liteSignature offchain verification is implemented for now)
+  if (preVerifySignatureOffchain) {
+    const isValidSignature = priceSigner.verifyLiteSignature({
+      pricePackage: {
+        prices: fetchedPackage.prices,
+        timestamp: fetchedPackage.timestamp,
+      },
+      signer,
+      signature: fetchedPackage.signature,
```
Comment (jakub-wojciechowski) [inline utils/v2/connector/impl/DataPackageUtils.ts:63]: Do we need a signature for light signature verification?

File: utils/v2/connector/impl/DataPackageUtils.ts (hunk 1 above)
Comment (hatskier) [reply utils/v2/connector/impl/DataPackageUtils.ts:22]: `first-valid` will work in a completely different way than `newest-valid`. It will update the data fetching logic and will not wait for all responses. Instead, it will listen on each response and will ise the first valid one. It allows to minimise time of data fetching and can be used for dapps that need to have data with the lowest possible latency.

`newest-valid` waits for all responses and selects the valid one with the biggest timestamp

File: utils/v2/connector/impl/DataPackageUtils.ts (hunk 2 above)
Comment (hatskier) [reply utils/v2/connector/impl/DataPackageUtils.ts:63]: Nope, but the function receives an object of type, that must have this prop :)

File: utils/v2/connector/impl/RedStonePriceFeed.ts [hunk 3]
```diff
@@ -1,47 +1,116 @@
...
+  signers?: string[], // Will be used in future (after multi-nodes creation)
+  sources: SourceConfig[],
+};
+
+export interface PriceFeedOptions {
+  dataSources?: DataSourcesConfig;
+  asset?: string;
+};
+
+export type RedStoneProvider =
```
Comment (jakub-wojciechowski) [inline utils/v2/connector/impl/RedStonePriceFeed.ts:27]: Do you think it's good to strong-type it here, wouldn't it be harder to add new providers?

File: utils/v2/connector/impl/RedStonePriceFeed.ts [hunk 4]
```diff
@@ -1,47 +1,116 @@
...
+      || DEFAULT_TIMEOUT_MILLISECONDS;
+    const promises = this.fetchers.map(fetcher =>
+      fetcher.getLatestDataWithTimeout(this.providerId, timeoutMilliseconds));
+    const results = await Promise.allSettled(promises);
+
+    // Validating fetched data
+    const fulfilledPromisesResults = results.filter(r => r.status === "fulfilled") as
+      PromiseFulfilledResult<SignedDataPackageResponse>[];
+    const dataPackages = fulfilledPromisesResults.map(r => r.value);
+    const validDataPackages = dataPackages.filter(p =>
```
Comment (jakub-wojciechowski) [inline utils/v2/connector/impl/RedStonePriceFeed.ts:81]: How expensive is validation?
An alternative approach will be to first optimistically select the package (latest) and validate only a single package.

File: utils/v2/connector/impl/RedStonePriceFeed.ts [hunk 5]
```diff
@@ -1,47 +1,116 @@
...
 
-    return {
-      priceData,
-      ..._.pick(response.data, ["signer", "signature", "liteSignature"]),
-    };
+    return convertResponseToPricePackage(finalResponse, this.cachedSigner!);
   }
 
+  // TODO: get rid of it later
+  // It is a potential single point of failure
```
Comment (jakub-wojciechowski) [inline utils/v2/connector/impl/RedStonePriceFeed.ts:103]: But the signer is only needed for a pre-validation. 
We could always skip it and rely only on the on-chain validation

File: utils/v2/connector/impl/fetchers/StreamrFetcher.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    }
+    return this.lastValue;
+  }
+
+  protected getStreamId() {
+    return `${this.config.streamrEndpointPrefix!}/`
+      + (this.asset ? "prices" : "package");
+  }
+
+  protected extractPriceValue(receivedValue: any): SignedDataPackageResponse {
```
Comment (jakub-wojciechowski) [inline utils/v2/connector/impl/fetchers/StreamrFetcher.ts:41]: Do we only stream single prices?

File: utils/v2/connector/impl/fetchers/Fetcher.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+export interface SourceConfig {
+  type: SourceType;
+  url?: string; // required for "cache-layer" sources
+  streamrEndpointPrefix?: string; // required for "streamr" and "streamr-historical" sources
+  disabledForSinglePrices?: boolean;
+}
+
+export abstract class Fetcher {
+  constructor(protected config: SourceConfig, protected asset?: string) {}
```
Comment (jakub-wojciechowski) [inline utils/v2/connector/impl/fetchers/Fetcher.ts:20]: If the asset is not defined, we fetch the whole package, right?
Maybe we should comment on it, so it's easier to understand.

File: utils/v2/connector/impl/RedStonePriceFeed.ts (hunk 3 above)
Comment (hatskier) [reply utils/v2/connector/impl/RedStonePriceFeed.ts:27]: I would keep it this way for now, as we actually want to motivate people to collaborate with us while using oracles. Because we want to avoid situations, when someone uses our oracles (in a wrong way) and tells that redstone doesn't work

File: utils/v2/connector/impl/RedStonePriceFeed.ts (hunk 4 above)
Comment (hatskier) [reply utils/v2/connector/impl/RedStonePriceFeed.ts:81]: I think, it's not really expensive, but will do performance testing

File: utils/v2/connector/impl/RedStonePriceFeed.ts [hunk 8]
```diff
@@ -1,47 +1,116 @@
...
+      // TODO: get rid of it, it's a potential single point of failure
+      this.getSigner(); // we are loading signer public key in advance
   }
 
+  // This is the entrypoint function of this module
   async getSignedPrice(): Promise<SignedPriceDataType> {
-    const response = await axios.get(this.apiUrl);
+    // We need to get signer public key firstly
+    // It will be used to pre-verify signatures off-chain
+    await this.getSigner();
```
Comment (jakub-wojciechowski) [inline utils/v2/connector/impl/RedStonePriceFeed.ts:67]: How about doing it in a lazy way - so we only fetch it in the pre-validation mode. 
(We could skip one request in case of a low-latency approach)

File: utils/v2/connector/impl/RedStonePriceFeed.ts (hunk 5 above)
Comment (hatskier) [reply utils/v2/connector/impl/RedStonePriceFeed.ts:103]: That's true, but if we use many sources, it would make sense to validate it before sending potentially invalid transaction. Of course it depends on the time costs of offchain signature verification. So I'll do the performance tests

File: utils/v2/connector/impl/fetchers/StreamrFetcher.ts (hunk 6 above)
Comment (hatskier) [reply utils/v2/connector/impl/fetchers/StreamrFetcher.ts:41]: No. Why do you think so?

File: utils/v2/connector/impl/fetchers/Fetcher.ts (hunk 7 above)
Comment (hatskier) [reply utils/v2/connector/impl/fetchers/Fetcher.ts:20]: Right. Great idea!

File: utils/v2/connector/impl/RedStonePriceFeed.ts [hunk 9]
```diff
@@ -1,47 +1,116 @@
...
+    
+    // Fetching data from all sources simultaneously with timeout
+    const timeoutMilliseconds =
+      this.priceFeedOptions.dataSources?.timeoutMilliseconds
+      || DEFAULT_TIMEOUT_MILLISECONDS;
+    const promises = this.fetchers.map(fetcher =>
+      fetcher.getLatestDataWithTimeout(this.providerId, timeoutMilliseconds));
+    const results = await Promise.allSettled(promises);
```
Comment (jakub-wojciechowski) [inline utils/v2/connector/impl/RedStonePriceFeed.ts:75]: Does it mean that we should wait for all the promises to resolve or timeout?
I'm guessing the most common way will be to get the first answer, shouldn't we resolve after the first fetcher returns the value?

File: utils/v2/connector/impl/RedStonePriceFeed.ts (hunk 8 above)
Comment (hatskier) [reply utils/v2/connector/impl/RedStonePriceFeed.ts:67]: We should update this logic anyway if we want to support multiple signers. But I agree that it's better not to do things that are not required :)

File: utils/v2/connector/impl/RedStonePriceFeed.ts (hunk 9 above)
Comment (hatskier) [reply utils/v2/connector/impl/RedStonePriceFeed.ts:75]: I agree, I've told about it in the first comment about `first-valid` and `newest-valid` :)

File: utils/v2/connector/impl/DataPackageUtils.ts (hunk 1 above)
Comment (jakub-wojciechowski) [reply utils/v2/connector/impl/DataPackageUtils.ts:22]: But is it implemented yet?

File: utils/v2/connector/impl/RedStonePriceFeed.ts (hunk 3 above)
Comment (jakub-wojciechowski) [reply utils/v2/connector/impl/RedStonePriceFeed.ts:27]: Fine, but could we keep somewhere a document what's necessary to add a new provider, so we won't forget that we should update the evm-connector code.

File: utils/v2/connector/impl/RedStonePriceFeed.ts (hunk 8 above)
Comment (jakub-wojciechowski) [reply utils/v2/connector/impl/RedStonePriceFeed.ts:67]: Ok, that should be a small change
(We will also use the getter with a variable caching an answer). Let's do it not to forget

File: utils/v2/connector/impl/RedStonePriceFeed.ts (hunk 4 above)
Comment (jakub-wojciechowski) [reply utils/v2/connector/impl/RedStonePriceFeed.ts:81]: Hmm, I think it's more important to implement "first-valid" approach first. 
I'm really afraid to be bound by the worst/slowest fetcher every time.

File: utils/v2/connector/impl/fetchers/StreamrFetcher.ts (hunk 6 above)
Comment (jakub-wojciechowski) [reply utils/v2/connector/impl/fetchers/StreamrFetcher.ts:41]: My mistake, I haven't noticed the else part

File: utils/v2/connector/impl/DataPackageUtils.ts (hunk 1 above)
Comment (hatskier) [reply utils/v2/connector/impl/DataPackageUtils.ts:22]: No, I wanted to implement it after adding few mirror nodes. But can do it now also

File: utils/v2/connector/impl/RedStonePriceFeed.ts (hunk 3 above)
Comment (hatskier) [reply utils/v2/connector/impl/RedStonePriceFeed.ts:27]: Sure! However I would not forget about it, because I would do an end to end testing each time when we add a new provider :)

File: utils/v2/connector/impl/RedStonePriceFeed.ts (hunk 4 above)
Comment (hatskier) [reply utils/v2/connector/impl/RedStonePriceFeed.ts:81]: Ok, I will implement it

File: test/message-based/PriceAware.streamr-storage.test.ts [hunk 10]
```diff
@@ -75,4 +75,54 @@ describe("Price Aware - streamr", function () {
       await syncTime(); // recommended for hardhat test
       await sample.executeWithPrice(toBytes32("AVAX"));
   });
+
+  it("should get price with single asset from streamr (first-valid)", async function () {
```
Comment (c-gma) [inline test/message-based/PriceAware.streamr-storage.test.ts:79]: Why do we only test a single dataSource (streamr-storage). 
Shouldn't we try to test at least 2 of them to check if it ends delivering the first valid price?

File: test/storage-based/RedstoneData.test.ts [hunk 11]
```diff
@@ -15,6 +15,10 @@ const toBytes32 = ethers.utils.formatBytes32String;
 const serialized = function (x: number): number {
   return x * 10 ** 8;
 };
+function checkPricesDeviation(value1: number, value2: number, maxPercentageDiff: number = 5) {
+  const percentDiff = 100 * Math.abs(1 - value1 / value2);
```
Comment (c-gma) [inline test/storage-based/RedstoneData.test.ts:19]: Doesn't it depend on the ordering of values. For example lest take 100 and 10
1) 100 * abs(1-10) = 900
2) 100 * abs(1-0.1) = 90

File: test/storage-based/RedstoneDataWithClearing.test.ts [hunk 12]
```diff
@@ -18,6 +18,11 @@ const serialized = function (x: number): number {
   return x * 10 ** 8;
 };
 
+function checkPricesDeviation(value1: number, value2: number, maxPercentageDiff: number = 5) {
```
Comment (c-gma) [inline test/storage-based/RedstoneDataWithClearing.test.ts:21]: duplicated code, let's move it to utils.

File: test/storage-based/RedstoneData.test.ts (hunk 11 above)
Comment (c-gma) [reply test/storage-based/RedstoneData.test.ts:19]: I think it could be good enough to treat one number as a base value and document it.

File: utils/v2/connector/impl/RedStonePriceFeed.ts [hunk 13]
```diff
@@ -65,11 +66,45 @@ export class RedStonePriceFeed implements PriceFeedConnector {
...
+      return (async () => {
+        const response = await fetcher.getLatestDataWithTimeout(
+          this.providerId,
+          timeoutMilliseconds);
+        const isValid = validateDataPackage(response, this.priceFeedOptions, this.cachedSigner!);
+        if (isValid) {
+          return response;
+        } else {
+          console.warn("Invalid response: " + JSON.stringify(response));
+          throw new Error(
```
Comment (c-gma) [inline utils/v2/connector/impl/RedStonePriceFeed.ts:96]: Shouldn't we pass fetcherIndex as an argument to the async() method?

File: utils/v2/connector/impl/RedStonePriceFeed.ts [hunk 14]
```diff
@@ -1,5 +1,6 @@
 import { PriceFeedConnector, SignedPriceDataType } from "../PriceFeedConnector";
 import axios from "axios";
+import bluebird from "bluebird";
```
Comment (c-gma) [inline utils/v2/connector/impl/RedStonePriceFeed.ts:3]: Why are you using bluebird, is it better than Promise from ES2021 (https://javascript.plainenglish.io/promise-any-and-promise-race-in-es2021-3250733b48eb) or is it not fully supported by browsers?

File: utils/v2/connector/impl/DataPackageUtils.ts (hunk 1 above)
Comment (hatskier) [reply utils/v2/connector/impl/DataPackageUtils.ts]: It's already implemented.

File: utils/v2/connector/impl/RedStonePriceFeed.ts (hunk 4 above)
Comment (hatskier) [reply utils/v2/connector/impl/RedStonePriceFeed.ts:114]: It's completed

File: utils/v2/connector/impl/RedStonePriceFeed.ts (hunk 14 above)
Comment (hatskier) [reply utils/v2/connector/impl/RedStonePriceFeed.ts:3]: Bluebird is more efficient and has features that are not available in standard JS/TS. E.g. promise cancellation, that can be added in future to redstone-evm-connector. I also had some traoubles with running ES2021 on my machine, so very probably it can case similar problems with compactibility on other devices

File: test/message-based/PriceAware.streamr-storage.test.ts (hunk 10 above)
Comment (hatskier) [reply test/message-based/PriceAware.streamr-storage.test.ts:83]: We test few of them in other tests, where we use default data Source config for redstone-avalanche price feed. It uses many sources. I've also added similar tests to `PriceAwareWithManySigners.test.ts`

File: utils/v2/connector/impl/RedStonePriceFeed.ts (hunk 13 above)
Comment (hatskier) [reply utils/v2/connector/impl/RedStonePriceFeed.ts:89]: I am not sure I understand the question. Can you please provide more details?

File: test/storage-based/RedstoneData.test.ts (hunk 11 above)
Comment (hatskier) [reply test/storage-based/RedstoneData.test.ts:19]: Fixed

File: test/storage-based/RedstoneDataWithClearing.test.ts (hunk 12 above)
Comment (hatskier) [reply test/storage-based/RedstoneDataWithClearing.test.ts:21]: Done

File: utils/v2/connector/impl/fetchers/Fetcher.ts (hunk 7 above)
Comment (hatskier) [reply utils/v2/connector/impl/fetchers/Fetcher.ts:24]: Added the comment

File: utils/v2/connector/impl/RedStonePriceFeed.ts (hunk 8 above)
Comment (hatskier) [reply utils/v2/connector/impl/RedStonePriceFeed.ts]: Did it

File: contracts/message-based/PriceAware.sol [hunk 15]
```diff
@@ -26,6 +26,10 @@ contract PriceAware is Ownable {
     emit TrustedSignerChanged(trustedSigner);
   }
 
+  function isSignerAuthorized(address _receviedSigner) internal virtual view returns (bool) {
```
Comment (jakub-wojciechowski) [inline contracts/message-based/PriceAware.sol:29]: If we're adding this method we may remove the getTrustedSigner one to reduce the complexity and gas costs:
_receviedSigner == trustedSigner

File: test/message-based/PriceAwareWithManySigners.test.ts [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+                    valueSelectionAlgorithm: "first-valid",
+                    timeoutMilliseconds: 10000,
+                    maxTimestampDiffMilliseconds: 150000,
+                    preVerifySignatureOffchain: false,
+                },
+            });
+    
+        // await sample.authorizeProvider(); <- note, it's not reuired in this test
+        await syncTime(); // recommended for hardhat test
+        await sample.executeWithPrice(toBytes32("AVAX"));
```
Comment (jakub-wojciechowski) [inline test/message-based/PriceAwareWithManySigners.test.ts:122]: How would it behave if all providers fail?
(Would it throw exception, will we catch it)
[It'll be great to also test the negative scenario]

File: utils/v2/connector/impl/RedStonePriceFeed.ts (hunk 13 above)
Comment (jakub-wojciechowski) [reply utils/v2/connector/impl/RedStonePriceFeed.ts:89]: The idea was to check if we shouldn't wrap the argument and connect it to promise's context:
https://stackoverflow.com/questions/35318442/how-to-pass-parameter-to-a-promise-function

to be sure that changing promise will read the fetcherIndex with value at the time it's passed to the promise.

Comment (hatskier) [reply utils/v2/connector/impl/RedStonePriceFeed.ts:89]: Oh, right sorry, just fixed. Thank you for noticing!

File: contracts/message-based/PriceAware.sol (hunk 15 above)
Comment (hatskier) [reply contracts/message-based/PriceAware.sol:29]: I would leave it to be compatible with our previous approaches. If we override the `isSignerAuthorized` function the `getTrustedSigner` should not even be called so it shouldn't consume a lot of gas for just existing.

File: test/message-based/PriceAwareWithManySigners.test.ts (hunk 16 above)
Comment (hatskier) [reply test/message-based/PriceAwareWithManySigners.test.ts:123]: Good idea. Added few tests for bad cases
